File: csharp/RobustMQ.Mq9/MQ9Client.cs
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using NATS.Client.Core;

namespace RobustMQ.Mq9;

/// <summary>
/// Async mq9 client for RobustMQ.
/// </summary>
/// <example>
/// <code>
/// await using var client = new MQ9Client("nats://localhost:4222");
/// await client.ConnectAsync();
///
/// var mailbox = await client.CreateAsync(ttl: 3600);
/// await client.SendAsync(mailbox.MailId, "hello"u8.ToArray());
///
/// await client.SubscribeAsync(mailbox.MailId, async msg =>
/// {
///     Console.WriteLine(Encoding.UTF8.GetString(msg.Payload));
/// });
/// </code>
/// </example>
public class MQ9Client : IAsyncDisposable
{
    private const string Prefix = "$mq9.AI";
    private const string MailboxCreate = $"{Prefix}.MAILBOX.CREATE";
    private static readonly TimeSpan RequestTimeout = TimeSpan.FromSeconds(5);

    private readonly string _server;
    // Internal for testing — allows injecting a mock transport
    internal INatsConnection? _nc;

    public MQ9Client(string server = "nats://localhost:4222")
    {
        _server = server;
    }

    // ── Connection ────────────────────────────────────────────────────────────

    public async Task ConnectAsync()
    {
        var opts = new NatsOpts { Url = _server };
        var nc = new NatsConnection(opts);
        await nc.ConnectAsync();
        _nc = nc;
    }

    public async ValueTask DisposeAsync()
    {
        if (_nc is NatsConnection nc)
            await nc.DisposeAsync();
    }

    // ── Create ────────────────────────────────────────────────────────────────

    public async Task<Mailbox> CreateAsync(int ttl, bool isPublic = false,
        string name = "", string desc = "")
    {
        if (isPublic && string.IsNullOrEmpty(name))
            throw new MQ9Error("name is required when public=true");

        var req = new JsonObject { ["ttl"] = ttl };
        if (isPublic)
        {
            req["public"] = true;
            req["name"] = name;
            if (!string.IsNullOrEmpty(desc)) req["desc"] = desc;
        }

        var resp = await RequestAsync(MailboxCreate, req);
        var mailId = resp["mail_id"]?.GetValue<string>()
                     ?? throw new MQ9Error("invalid response: missing mail_id");
        return new Mailbox(mailId, isPublic, name, desc);
    }

    // ── Send ──────────────────────────────────────────────────────────────────

    public async Task SendAsync(string mailId, byte[] payload,
        Priority priority = Priority.Normal)
    {
        EnsureConnected();
        var subject = $"{Prefix}.MAILBOX.MSG.{mailId}.{priority.ToSubjectString()}";
        await _nc!.PublishAsync(subject, payload);
    }

    // ── Subscribe ─────────────────────────────────────────────────────────────

    public async Task<IAsyncDisposable> SubscribeAsync(
        string mailId,
        Func<Mq9Message, Task> callback,
        Priority? priority = null,
        string queueGroup = "")
    {
        EnsureConnected();
        var subject = priority.HasValue
            ? $"{Prefix}.MAILBOX.MSG.{mailId}.{priority.Value.ToSubjectString()}"
            : $"{Prefix}.MAILBOX.MSG.{mailId}.*";

        var sub = string.IsNullOrEmpty(queueGroup)
            ? await _nc!.SubscribeCoreAsync<byte[]>(subject)
            : await _nc!.SubscribeCoreAsync<byte[]>(subject, queueGroup: queueGroup);

        // Start background consumer
        var cts = new CancellationTokenSource();
        var task = Task.Run(async () =>
        {
            await foreach (var msg in sub.Msgs.ReadAllAsync(cts.Token))
            {
                try
                {
                    var m = ParseIncoming(mailId, msg.Subject, msg.Data ?? Array.Empty<byte>());
                    await callback(m);
                }
                catch
                {
                    // swallow callback exceptions to keep subscription alive
                }
            }
        }, cts.Token);

        return new SubscriptionHandle(sub, cts);
    }

    // ── List ──────────────────────────────────────────────────────────────────

    public async Task<List<MessageMeta>> ListAsync(string mailId)
    {
        var subject = $"{Prefix}.MAILBOX.LIST.{mailId}";
        var resp = await RequestAsync(subject, new JsonObject());
        var metas = new List<MessageMeta>();
        var arr = resp["messages"]?.AsArray();
        if (arr != null)
            foreach (var node in arr)
                if (node != null)
                    metas.Add(new MessageMeta(
                        node["msg_id"]?.GetValue<string>() ?? "",
                        PriorityExtensions.FromString(node["priority"]?.GetValue<string>() ?? "normal"),
                        node["ts"]?.GetValue<long>() ?? 0));
        return metas;
    }

    // ── Delete ────────────────────────────────────────────────────────────────

    public async Task DeleteAsync(string mailId, string msgId)
    {
        var subject = $"{Prefix}.MAILBOX.DELETE.{mailId}.{msgId}";
        await RequestAsync(subject, new JsonObject());
    }

    // ── Internal helpers ─────────────────────────────────────────────────────

    private void EnsureConnected()
    {
        if (_nc == null)
            throw new InvalidOperationException("Not connected. Call ConnectAsync() first.");
    }

    private async Task<JsonNode> RequestAsync(string subject, JsonNode payload)
    {
        EnsureConnected();
        var data = Encoding.UTF8.GetBytes(payload.ToJsonString());

        using var cts = new CancellationTokenSource(RequestTimeout);
        var reply = await _nc!.RequestAsync<byte[], byte[]>(subject, data, cancellationToken: cts.Token);

        if (reply.Data == null)
            throw new MQ9Error($"empty response from {subject}");

        var node = JsonNode.Parse(reply.Data)
                   ?? throw new MQ9Error("invalid JSON response");

        if (node["error"] != null)
        {
            var code = node["code"]?.GetValue<int>() ?? 0;
            throw new MQ9Error(node["error"]!.GetValue<string>(), code);
        }
        return node;
    }

    private static Mq9Message ParseIncoming(string mailId, string subject, byte[] data)
    {
        // Extract priority from last subject token
        var parts = subject.Split('.');
        var priorityStr = parts.Length > 0 ? parts[^1] : "normal";
        var priority = PriorityExtensions.FromString(priorityStr);

        try
        {
            var node = JsonNode.Parse(data);
            if (node?["msg_id"] != null)
                return ParseMessageNode(mailId, node);
        }
        catch { /* fall through */ }

        return new Mq9Message("", mailId, priority, data);
    }

    private static Mq9Message ParseMessageNode(string mailId, JsonNode node)
    {
        var msgId = node["msg_id"]?.GetValue<string>() ?? "";
        var priority = PriorityExtensions.FromString(node["priority"]?.GetValue<string>() ?? "normal");
        var rawPayload = node["payload"]?.GetValue<string>() ?? "";
        var payload = string.IsNullOrEmpty(rawPayload)
            ? Array.Empty<byte>()
            : Convert.FromBase64String(rawPayload);
        return new Mq9Message(msgId, mailId, priority, payload);
    }
}

// ── Subscription handle ───────────────────────────────────────────────────────

file sealed class SubscriptionHandle : IAsyncDisposable
{
    private readonly IDisposable _sub;
    private readonly CancellationTokenSource _cts;

    public SubscriptionHandle(IDisposable sub, CancellationTokenSource cts)
    {
        _sub = sub;
        _cts = cts;
    }

    public ValueTask DisposeAsync()
    {
        _cts.Cancel();
        _sub.Dispose();
        _cts.Dispose();
        return ValueTask.CompletedTask;
    }
}

File: java/src/main/java/com/robustmq/mq9/MQ9Client.java
package com.robustmq.mq9;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.nats.client.Connection;
import io.nats.client.Dispatcher;
import io.nats.client.Message;
import io.nats.client.Nats;
import io.nats.client.Options;

import java.io.IOException;
import java.time.Duration;
import java.util.ArrayList;
import java.util.Base64;
import java.util.List;
import java.util.concurrent.CompletableFuture;

/**
 * Async mq9 client for RobustMQ.
 *
 * <pre>{@code
 * MQ9Client client = new MQ9Client("nats://localhost:4222");
 * client.connect();
 *
 * Mailbox mailbox = client.create(3600).get();
 * client.send(mailbox.getMailId(), "hello".getBytes(), Priority.NORMAL).get();
 *
 * client.subscribe(mailbox.getMailId(), msg ->
 *     System.out.println(new String(msg.getPayload())),
 *     null, "").get();
 *
 * client.close();
 * }</pre>
 */
public class MQ9Client implements AutoCloseable {

    private static final String PREFIX = "$mq9.AI";
    private static final String MAILBOX_CREATE = PREFIX + ".MAILBOX.CREATE";
    private static final Duration REQUEST_TIMEOUT = Duration.ofSeconds(5);

    private final String server;
    private final ObjectMapper mapper = new ObjectMapper();
    // Package-private for testing: allows injecting a mock Connection
    Connection nc;

    public MQ9Client(String server) {
        this.server = server;
    }

    // ── Connection ────────────────────────────────────────────────────────────

    public void connect() throws MQ9Error {
        try {
            Options opts = new Options.Builder()
                    .server(server)
                    .maxReconnects(-1)
                    .build();
            nc = Nats.connect(opts);
        } catch (IOException | InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new MQ9Error("connection failed: " + e.getMessage());
        }
    }

    @Override
    public void close() {
        if (nc != null) {
            try {
                nc.drain(Duration.ofSeconds(5)).get();
            } catch (Exception e) {
                // best-effort drain
            }
        }
    }

    // ── Create ────────────────────────────────────────────────────────────────

    public CompletableFuture<com.robustmq.mq9.Mailbox> create(int ttl) {
        return createInternal(ttl, false, "", "");
    }

    public CompletableFuture<com.robustmq.mq9.Mailbox> createPublic(int ttl, String name, String desc) {
        if (name == null || name.isEmpty()) {
            return CompletableFuture.failedFuture(new MQ9Error("name is required when public=true"));
        }
        return createInternal(ttl, true, name, desc != null ? desc : "");
    }

    private CompletableFuture<com.robustmq.mq9.Mailbox> createInternal(int ttl, boolean isPublic,
                                                                        String name, String desc) {
        return CompletableFuture.supplyAsync(() -> {
            ObjectNode req = mapper.createObjectNode();
            req.put("ttl", ttl);
            if (isPublic) {
                req.put("public", true);
                req.put("name", name);
                if (!desc.isEmpty()) req.put("desc", desc);
            }
            JsonNode resp = doRequest(MAILBOX_CREATE, req);
            String mailId = resp.get("mail_id").asText();
            return new com.robustmq.mq9.Mailbox(mailId, isPublic, name, desc);
        });
    }

    // ── Send ──────────────────────────────────────────────────────────────────

    public CompletableFuture<Void> send(String mailId, byte[] payload, Priority priority) {
        return CompletableFuture.runAsync(() -> {
            ensureConnected();
            String subject = PREFIX + ".MAILBOX.MSG." + mailId + "." + priority.getValue();
            try {
                nc.publish(subject, payload);
            } catch (Exception e) {
                throw new RuntimeException(new MQ9Error("publish failed: " + e.getMessage()));
            }
        });
    }

    // ── Subscribe ─────────────────────────────────────────────────────────────

    public CompletableFuture<Dispatcher> subscribe(String mailId, MessageHandler handler,
                                                    Priority priority, String queueGroup) {
        return CompletableFuture.supplyAsync(() -> {
            ensureConnected();
            String sub = (priority != null)
                    ? PREFIX + ".MAILBOX.MSG." + mailId + "." + priority.getValue()
                    : PREFIX + ".MAILBOX.MSG." + mailId + ".*";

            Dispatcher dispatcher = nc.createDispatcher((Message msg) -> {
                com.robustmq.mq9.Message m = parseIncoming(mailId, msg);
                handler.onMessage(m);
            });

            if (queueGroup != null && !queueGroup.isEmpty()) {
                dispatcher.subscribe(sub, queueGroup);
            } else {
                dispatcher.subscribe(sub);
            }
            return dispatcher;
        });
    }

    // ── List ──────────────────────────────────────────────────────────────────

    public CompletableFuture<List<MessageMeta>> list(String mailId) {
        return CompletableFuture.supplyAsync(() -> {
            String subject = PREFIX + ".MAILBOX.LIST." + mailId;
            JsonNode resp = doRequest(subject, mapper.createObjectNode());
            List<MessageMeta> metas = new ArrayList<>();
            JsonNode arr = resp.get("messages");
            if (arr != null && arr.isArray()) {
                for (JsonNode node : arr) {
                    metas.add(new MessageMeta(
                            node.path("msg_id").asText(""),
                            Priority.fromString(node.path("priority").asText("normal")),
                            node.path("ts").asLong(0)
                    ));
                }
            }
            return metas;
        });
    }

    // ── Delete ────────────────────────────────────────────────────────────────

    public CompletableFuture<Void> delete(String mailId, String msgId) {
        return CompletableFuture.runAsync(() -> {
            String subject = PREFIX + ".MAILBOX.DELETE." + mailId + "." + msgId;
            doRequest(subject, mapper.createObjectNode());
        });
    }

    // ── Internal helpers ─────────────────────────────────────────────────────

    private void ensureConnected() {
        if (nc == null) {
            throw new IllegalStateException("Not connected. Call connect() first.");
        }
    }

    private JsonNode doRequest(String subject, ObjectNode payload) {
        ensureConnected();
        try {
            byte[] data = mapper.writeValueAsBytes(payload);
            io.nats.client.Message reply = nc.request(subject, data, REQUEST_TIMEOUT);
            if (reply == null) {
                throw new RuntimeException(new MQ9Error("request timed out: " + subject));
            }
            JsonNode resp = mapper.readTree(reply.getData());
            if (resp.has("error")) {
                int code = resp.path("code").asInt(0);
                throw new RuntimeException(new MQ9Error(resp.get("error").asText(), code));
            }
            return resp;
        } catch (RuntimeException e) {
            throw e;
        } catch (Exception e) {
            throw new RuntimeException(new MQ9Error("request failed: " + e.getMessage()));
        }
    }

    private com.robustmq.mq9.Message parseIncoming(String mailId, Message raw) {
        // Extract priority from subject last token
        String subject = raw.getSubject();
        String[] parts = subject.split("\\.");
        String priorityStr = parts.length > 0 ? parts[parts.length - 1] : "normal";
        Priority priority = Priority.fromString(priorityStr);

        // Try JSON envelope first
        try {
            JsonNode node = mapper.readTree(raw.getData());
            if (node.has("msg_id")) {
                return parseMessageNode(mailId, node);
            }
        } catch (Exception ignored) {
            // fall through to raw payload
        }

        return new com.robustmq.mq9.Message("", mailId, priority, raw.getData());
    }

    private com.robustmq.mq9.Message parseMessageNode(String mailId, JsonNode node) {
        String msgId = node.path("msg_id").asText("");
        Priority priority = Priority.fromString(node.path("priority").asText("normal"));
        String rawPayload = node.path("payload").asText("");
        byte[] payload = rawPayload.isEmpty() ? new byte[0] : Base64.getDecoder().decode(rawPayload);
        return new com.robustmq.mq9.Message(msgId, mailId, priority, payload);
    }
}
